### PR TITLE
Allow parallel parachain validation

### LINKED ISSUE #361: Allow parallel parachain validation
Currently out-of-process parachain executor only allows one process and one validation at a time. 
It should be possible to validate multiple parachains at once. Each parachain should be assigned a process so that copying the code is minimized.

### PR DESCRIPTION
**Pull request by arkpar**
Closes #361

### CODE REVIEW

File: parachain/src/wasm_executor/validation_host.rs [hunk 1]
```diff
@@ -40,7 +43,7 @@ enum Event {
 }
 
 lazy_static::lazy_static! {
-	pub static ref HOST: Mutex<ValidationHost> = Mutex::new(ValidationHost::new());
+	static ref HOSTS: Mutex<HashMap<ParaId, Arc<Mutex<ValidationHost>>>> = Mutex::new(Default::default());
```
Comment (rphmeier) [inline parachain/src/wasm_executor/validation_host.rs]: why organize it by para-id? i think it's more likely that we would run jobs of the same parachain at the same time than to run jobs of different parachains at the same time.

File: parachain/src/wasm_executor.rs [hunk 2]
```diff
@@ -27,10 +27,12 @@ use wasmi::{
 	ModuleImportResolver, RuntimeValue, Externals, Error as WasmError, ValueType,
 	memory_units::{self, Bytes, Pages, RoundUpTo}
 };
-use super::{ValidationParams, ValidationResult, MessageRef, UpwardMessageRef, UpwardMessage, IncomingMessage};
+use super::{ValidationParams, ValidationResult, MessageRef, UpwardMessageRef,
+	UpwardMessage, IncomingMessage, Id as ParaId};
```
Comment (bkchr) [inline parachain/src/wasm_executor.rs]:
```suggestion
	UpwardMessage, IncomingMessage, Id as ParaId
};
```

File: parachain/src/wasm_executor.rs [hunk 3]
```diff
@@ -27,10 +27,12 @@ use wasmi::{
 	ModuleImportResolver, RuntimeValue, Externals, Error as WasmError, ValueType,
 	memory_units::{self, Bytes, Pages, RoundUpTo}
 };
-use super::{ValidationParams, ValidationResult, MessageRef, UpwardMessageRef, UpwardMessage, IncomingMessage};
+use super::{ValidationParams, ValidationResult, MessageRef, UpwardMessageRef,
```
Comment (bkchr) [inline parachain/src/wasm_executor.rs]:
```suggestion
use super::{
	ValidationParams, ValidationResult, MessageRef, UpwardMessageRef,
```

File: parachain/src/wasm_executor/validation_host.rs (hunk 1 above)
Comment (bkchr) [reply parachain/src/wasm_executor/validation_host.rs]: And couldn't we drop the `Arc`?

Comment (arkpar) [reply parachain/src/wasm_executor/validation_host.rs]: The `Arc` is there so that each `ValidationHost` can be locked outside of the `HOSTS` mutex

**Comment by gavofyork** [comment]
looks like it'll need a bit of work to get building again.

**Pull request merged**